### PR TITLE
Fix reparenting subsets between rotation orientations

### PR DESCRIPTION
**Pull request by rosteen**
Previously, the subset reparenting didn't properly handle the rotation between two different orientations. This fixes it. I'll add test coverage tomorrow.

### CODE REVIEW

**Comment by pllim** [comment]
You would have to also remove the xfail here:

https://github.com/spacetelescope/jdaviz/blob/20de93b7fd7e5689da441be1f5cef2f9cf6bd8db/jdaviz/configs/imviz/tests/test_delete_data.py#L79

File: jdaviz/app.py [hunk 1]
```diff
@@ -1796,6 +1798,14 @@ def _reparent_subsets(self, old_parent, new_parent=None):
                     # Translate bounds through WCS if needed
                     if (self.config == "imviz" and
                             self._jdaviz_helper.plugins["Orientation"].link_type == "WCS"):
+
+                        old_angle, _,  old_flip = get_compass_info(old_parent.coords, (10, 10))[-3:]
+                        new_angle, _, new_flip = get_compass_info(new_parent.coords, (10, 10))[-3:]
+                        relative_angle = new_angle - old_angle
+                        print(old_angle, new_angle, relative_angle)
```
Comment (pllim) [inline jdaviz/app.py:1805]:
```suggestion
```

File: jdaviz/app.py [hunk 2]
```diff
@@ -1993,9 +2009,14 @@ def set_data_visibility(self, viewer_reference, data_label, visible=True, replac
 
     def vue_data_item_remove(self, event):
 
+        print(event)
```
Comment (pllim) [inline jdaviz/app.py:2012]:
```suggestion
```

**Review by pllim — COMMENTED** [review]
Need a change log.

And remove leftover debugging stuff.

File: jdaviz/app.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import math
```
Comment (pllim) [inline jdaviz/app.py:1]:
```suggestion
```

File: jdaviz/app.py [hunk 4]
```diff
@@ -1796,6 +1798,14 @@ def _reparent_subsets(self, old_parent, new_parent=None):
                     # Translate bounds through WCS if needed
                     if (self.config == "imviz" and
                             self._jdaviz_helper.plugins["Orientation"].link_type == "WCS"):
+
+                        old_angle, _,  old_flip = get_compass_info(old_parent.coords, (10, 10))[-3:]
```
Comment (pllim) [inline jdaviz/app.py:1802]:
```suggestion
                        old_angle, _, old_flip = get_compass_info(old_parent.coords, (10, 10))[-3:]
```

**Review by pllim — COMMENTED** [review]
PEP 8

File: jdaviz/app.py (hunk 1 above)
Comment (rosteen) [reply jdaviz/app.py:1805]: Yes thanks, there's a few things to do, that's why it's in draft.

File: jdaviz/configs/imviz/tests/test_delete_data.py [hunk 5]
```diff
@@ -66,17 +65,16 @@ def test_delete_with_subset_wcs(self):
...
+        assert subset1.subset_state.xatt.parent.label == "Default orientation"
+        assert_allclose(subset1.subset_state.center(), (1.75, 1.75))
 
-        assert subset2.subset_state.xatt.parent.label == "has_wcs_2[SCI,1]"
-        assert_allclose(subset2.subset_state.roi.xmin, 1)
-        assert_allclose(subset2.subset_state.roi.ymin, 0, atol=1e-6)
-        assert_allclose(subset2.subset_state.roi.xmax, 3)
-        assert_allclose(subset2.subset_state.roi.ymax, 2)
+        assert subset2.subset_state.xatt.parent.label == "Default orientation"
+        assert_allclose(subset2.subset_state.roi.xmin, -0.25)
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_delete_data.py:72]: Hmm the 0.25 offset is still there.

File: jdaviz/configs/imviz/tests/test_delete_data.py [hunk 6]
```diff
@@ -66,17 +65,16 @@ def test_delete_with_subset_wcs(self):
...
+        assert_allclose(subset2.subset_state.roi.ymin, -0.25, atol=1e-6)
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_delete_data.py:73]: Do you still need this?
```suggestion
        assert_allclose(subset2.subset_state.roi.ymin, -0.25)
```

**Comment by pllim** [comment]
p.s. Looks like I am not off the hook from https://github.com/spacetelescope/jdaviz/issues/2724 since you didn't have to remove the AssertionError from these.

https://github.com/spacetelescope/jdaviz/blob/0e0a3b72bf5ccb55bf8d2068fa9da10e372f6fd7/jdaviz/configs/imviz/tests/test_orientation.py#L140-L145

File: jdaviz/configs/imviz/tests/test_delete_data.py (hunk 5 above)
Comment (rosteen) [reply jdaviz/configs/imviz/tests/test_delete_data.py:72]: Ah right, I hadn't been fully following the conversation about the 0.25 pixel issue, didn't think about it when I updated these. Seemed separate from the actual changes in this PR.

File: jdaviz/configs/imviz/tests/test_delete_data.py (hunk 6 above)
Comment (rosteen) [reply jdaviz/configs/imviz/tests/test_delete_data.py:73]: Probably not!

File: jdaviz/app.py [hunk 7]
```diff
@@ -1816,18 +1824,39 @@ def _reparent_subsets(self, old_parent, new_parent=None):
...
-                            roi.ymin = y_min
-                            roi.ymax = y_max
+                            x1, y1 = pixel_to_pixel(old_parent.coords, new_parent.coords,
+                                                    roi.xmin, roi.ymin)
+                            x2, y2 = pixel_to_pixel(old_parent.coords, new_parent.coords,
+                                                    roi.xmin, roi.ymax)
+                            x3, y3 = pixel_to_pixel(old_parent.coords, new_parent.coords,
+                                                    roi.xmax, roi.ymax)
+                            x3, y3 = pixel_to_pixel(old_parent.coords, new_parent.coords,
+                                                    roi.xmax, roi.ymin)
```
Comment (pllim) [inline jdaviz/app.py:1840]: `pixel_to_pixel` can vectorize, I think, see https://docs.astropy.org/en/stable/api/astropy.wcs.utils.pixel_to_pixel.html , so you really only need one call to that function.

File: jdaviz/app.py [hunk 8]
```diff
@@ -1995,7 +2024,11 @@ def vue_data_item_remove(self, event):
 
         data_label = event['item_name']
         data = self.data_collection[data_label]
-        self._reparent_subsets(data)
+        if self.config == "imviz":
```
Comment (pllim) [inline jdaviz/app.py:2027]: Also have to check link type here? You don't want this happening when linked by pixel, right? Or is the link type being handled somewhere else?

File: jdaviz/configs/imviz/tests/test_delete_data.py (hunk 5 above)
Comment (pllim) [reply jdaviz/configs/imviz/tests/test_delete_data.py:72]: Yeah, it is unrelated to your ticket. I am just thinking aloud. Sorry for the confusion.

File: jdaviz/app.py (hunk 7 above)
Comment (rosteen) [reply jdaviz/app.py]: Hmm, I tried calling it a couple different ways, and it doesn't seem like vectorizing analogously to `pixel_to_world` works for `pixel_to_pixel`. I did remove an extra call, realized I was setting then overwriting x3 and y3.

Comment (pllim) [reply jdaviz/app.py]: Oh, OK, maybe the API doc is misleading then. Thanks for trying!

File: jdaviz/app.py [hunk 9]
```diff
@@ -1816,18 +1824,37 @@ def _reparent_subsets(self, old_parent, new_parent=None):
...
+                            # New min/max before applying theta
+                            roi.xmin = new_center[0] - new_width/2
+                            roi.xmax = new_center[0] + new_width/2
+                            roi.ymin = new_center[1] - new_height/2
+                            roi.ymax = new_center[1] + new_height/2
+
+                        # Account for rotation between orientations
+                        if hasattr(roi, "theta"):
+                            angle = getattr(roi, "theta")
+                            setattr(roi, "theta", angle - np.deg2rad(relative_angle))
```
Comment (pllim) [inline jdaviz/app.py:1857]: You already check for `hasattr` above, can't you just access it directly like this? Am I missing something?
```suggestion
                            roi.theta = roi.theta - np.deg2rad(relative_angle)
```

File: jdaviz/app.py [hunk 10]
```diff
@@ -1796,11 +1797,18 @@ def _reparent_subsets(self, old_parent, new_parent=None):
                     # Translate bounds through WCS if needed
                     if (self.config == "imviz" and
                             self._jdaviz_helper.plugins["Orientation"].link_type == "WCS"):
+
+                        old_angle, _, old_flip = get_compass_info(old_parent.coords, (10, 10))[-3:]
```
Comment (pllim) [inline jdaviz/app.py:1801]: Do we want to add a comment that this shape is from here?

https://github.com/spacetelescope/jdaviz/blob/0e0a3b72bf5ccb55bf8d2068fa9da10e372f6fd7/jdaviz/configs/imviz/wcs_utils.py#L455

Comment (pllim) [reply jdaviz/app.py:1801]: Or better yet, programmatically accessing the shape rather than hardcoding?

Comment (rosteen) [reply jdaviz/app.py:1801]: I actually wasn't even sure if the shape matters at all, given that we only need the angle and flip. I can add a comment though,

File: jdaviz/app.py [hunk 11]
```diff
@@ -1796,11 +1797,18 @@ def _reparent_subsets(self, old_parent, new_parent=None):
...
+                        new_angle, _, new_flip = get_compass_info(new_parent.coords, (10, 10))[-3:]
+                        relative_angle = new_angle - old_angle
+                        if old_flip != new_flip:
+                            relative_angle += 180
```
Comment (kecnry) [inline jdaviz/app.py:1805]: this only matters for asymmetric shapes right?  In which case, rotating by 180 isn't the same as flipping, but maybe is better than nothing? 🤷

File: jdaviz/app.py (hunk 8 above)
Comment (kecnry) [reply jdaviz/app.py:2025]: Can we check something else other than the config?  Maybe check if the orientation plugin is loaded (which could probably be done in one line to avoid building helper.plugins multiple times, or by accessing the tray_items which is probably cheaper)?

File: jdaviz/app.py (hunk 11 above)
Comment (rosteen) [reply jdaviz/app.py:1805]: Hmm, I think you're right that this isn't the proper handling for the flipped case, thinking about it again I don't know what I thought a 180 degree rotation for a symmetric shape would do 😆 . I just tested going from E-left to E-right and deleting E-left, and the subset disappeared (which...implies something else is wrong). I'll do some more thinking about the best way to handle that case.

**Comment by rosteen** [comment]
@kecnry @pllim Most of the changes today were implementing Kyle's suggestion to check for the Orientation plugin, and fixing a semi-related bug I discovered while working on this (If you created a third orientation when a second one was active, then delete the second one, the third wouldn't be linked to anything). Should be ready for re-review tomorrow, I'll ping you then.

File: jdaviz/app.py (hunk 9 above)
Comment (rosteen) [reply jdaviz/app.py]: Did this locally and added you as a co-author.

**Comment by rosteen** [comment]
@kecnry @pllim This is ready for another look.

**Review by pllim — COMMENTED** [review]
Code looks like code. Some minor comments.

File: jdaviz/app.py [hunk 12]
```diff
@@ -1816,18 +1828,41 @@ def _reparent_subsets(self, old_parent, new_parent=None):
...
+                            roi.ymin = new_center[1] - new_height/2
+                            roi.ymax = new_center[1] + new_height/2
+
+                        # Account for rotation between orientations
+                        if hasattr(roi, "theta"):
+                            angle = getattr(roi, "theta")
+                            if old_flip != new_flip:
+                                new_angle = (np.deg2rad(relative_angle) - angle) % (2 * math.pi)
+                            else:
+                                new_angle = (angle - np.deg2rad(relative_angle)) % (2 * math.pi)
```
[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/app.py:1864]: Do we need a more involved test with these different flips?

**Comment by pllim** [comment]
I want to double check that the test catalog failure isn't transient or related. Gonna restart that job.

```
>       assert catalogs_plugin.number_of_results > 500
E       AssertionError: assert 0 > 500
E        +  where 0 = Catalogs...

jdaviz/configs/imviz/tests/test_catalogs.py:104: AssertionError
```

**Comment by pllim** [comment]
Nvm I see the same Catalogs failure on main. It is unrelated.

File: jdaviz/configs/imviz/tests/test_orientation.py [hunk 13]
```diff
@@ -152,3 +156,48 @@ def test_custom_orientation(self):
         lc_plugin._obj.add_orientation(rotation_angle=None, east_left=True, label=None,
                                        set_on_create=True, wrt_data=None)
         assert self.viewer.state.reference_data.label == "CCW 42.00 deg (E-left)"
+
+
+class TestDeleteOrientation(BaseImviz_WCS_WCS):
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_orientation.py:161]: Note: I made this its own test class so it does not pollute the other tests.

File: jdaviz/configs/imviz/tests/test_orientation.py [hunk 14]
```diff
@@ -152,3 +156,48 @@ def test_custom_orientation(self):
...
         assert self.viewer.state.reference_data.label == "CCW 42.00 deg (E-left)"
+
+
+class TestDeleteOrientation(BaseImviz_WCS_WCS):
+
+    @pytest.mark.parametrize("klass", [EllipseSkyRegion, RectangleSkyRegion])
+    @pytest.mark.parametrize(
+        ("angle", "sbst_theta"),
+        [(0.5 * u.rad, 2.641593),
+         (-0.5 * u.rad, 3.641589)])
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_orientation.py:167]: Is -0.5 rad a good input for "negative angle"? If not, feel free to adjust the test.

File: jdaviz/configs/imviz/tests/test_orientation.py [hunk 15]
```diff
@@ -152,3 +156,48 @@ def test_custom_orientation(self):
...
+    def test_delete_orientation_with_subset(self, klass, angle, sbst_theta):
+        lc_plugin = self.imviz.plugins['Orientation']
+        lc_plugin.link_type = 'WCS'
+
+        # Should automatically be applied as reference to first viewer.
+        lc_plugin._obj.create_north_up_east_left(set_on_create=True)
+
+        # Create rotated shape
+        reg = klass(center=SkyCoord(ra=337.51931488, dec=-20.83187472, unit="deg"),
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_orientation.py:176]: Providing region as input made more sense to me, so I did that change here. Also added test case for rectangle that seems to have a separate logic route. Luckily ellipse and rectangle have same inputs in regions, so we can easily parametrize them.

File: jdaviz/configs/imviz/tests/test_orientation.py [hunk 16]
```diff
@@ -152,3 +156,48 @@ def test_custom_orientation(self):
...
+        nuer_data = self.imviz.app.data_collection['North-up, East-right']
+        assert subset_group.subset_state.xatt in nuer_data.components
+        assert_allclose(subset_group.subset_state.roi.theta, sbst_theta, rtol=1e-5)
+
+        out_reg = self.imviz.app.get_subsets(include_sky_region=True)["Subset 1"][0]["sky_region"]
+        assert_allclose(out_reg.center.ra.deg, reg.center.ra.deg)
+        assert_allclose(out_reg.center.dec.deg, reg.center.dec.deg)
+        assert_quantity_allclose(out_reg.width, reg.width)
+        assert_quantity_allclose(out_reg.height, reg.height)
+        # FIXME: However, sky angle has to stay the same as per regions convention.
```
Comment (pllim) [inline jdaviz/configs/imviz/tests/test_orientation.py:201]: I think fixing this is out of scope for this PR. If you have a JIRA ticket from the "what does 0 angle mean" meeting today, please link here.

**Review by pllim — APPROVED** [review]
I pushed my remaining review as a commit, mainly expanding your new test. So this LGTM now. Thanks!

File: jdaviz/app.py [hunk 17]
```diff
@@ -1796,11 +1796,20 @@ def _reparent_subsets(self, old_parent, new_parent=None):
                     # Translate bounds through WCS if needed
                     if (self.config == "imviz" and
                             self._jdaviz_helper.plugins["Orientation"].link_type == "WCS"):
+
+                        # Default shape for WCS-only layers is 10x10, but it doesn't really matter
+                        # since we only need the angles.
+                        old_angle, _, old_flip = get_compass_info(old_parent.coords, (10, 10))[-3:]
+                        new_angle, _, new_flip = get_compass_info(new_parent.coords, (10, 10))[-3:]
+                        if old_flip != new_flip:
+                            relative_angle = 180 - new_angle - old_angle
```
Comment (kecnry) [inline jdaviz/app.py:1805]: I think this still assumes some symmetry in the subcomponent of the subset and would break if we ever introduce lasso subsets, right?  Maybe just worth a comment so we don't get tripped up by it in the future if we ever add that.

**Review by kecnry — APPROVED** [review]
Seems to work well!  Changing orientation and reparenting definitely isn't fast, so there may be room for optimization either here or elsewhere in the logic, but we can handle that separately.  Thanks for the fix!

File: jdaviz/app.py (hunk 17 above)
Comment (pllim) [reply jdaviz/app.py:1805]: FWIW, there is no lasso shape support in regions nor photutils. So it would probably be turned into some polygon that has no concept of angle anyway (though you would have to rotate it properly or whatever before you turn it into polygon), so maybe also put that in comment?

Comment (rosteen) [inline jdaviz/app.py:1805]:
```suggestion
                            # Note that this won't work for an irregular/assymetric region if we
                            # ever implement those.
                            relative_angle = 180 - new_angle - old_angle
```

**Comment by rosteen** [comment]
Thanks for the test updates @pllim, I added a comment about irregular regions and will merge after tests pass.
